File: Pharmaceutical_Company_Example/Creational/R/Factory_Method.R
# Product
RegulatoryDocument <- setRefClass(
  "RegulatoryDocument",
  fields = list(
    drug_name = "character",
    document_type = "character"
  ),
  methods = list(
    show_document = function() {
      cat("Regulatory Document for:", drug_name, "\n",
          "Type of Document: ", document_type, "\n\n")
    }
  )
)

# Abstract Creator
RegulatoryDocumentFactory <- setRefClass(
  "RegulatoryDocumentFactory",
  methods = list(
    create_document = function(drug_name) {
      stop("This method must be overridden by a concrete factory!")
    }
  )
)

# Concrete Creator for Vaccine Documents
VaccineDocumentFactory <- setRefClass(
  "VaccineDocumentFactory",
  contains = "RegulatoryDocumentFactory",
  methods = list(
    create_document = function(drug_name) {
      return(RegulatoryDocument$new(drug_name = drug_name, document_type = "Vaccine Regulatory Document"))
    }
  )
)

# Concrete Creator for Small-Molecule Documents
SmallMoleculeDocumentFactory <- setRefClass(
  "SmallMoleculeDocumentFactory",
  contains = "RegulatoryDocumentFactory",
  methods = list(
    create_document = function(drug_name) {
      return(RegulatoryDocument$new(drug_name = drug_name, document_type = "Small-Molecule Regulatory Document"))
    }
  )
)

# Concrete Creator for Biologics Documents
BiologicsDocumentFactory <- setRefClass(
  "BiologicsDocumentFactory",
  contains = "RegulatoryDocumentFactory",
  methods = list(
    create_document = function(drug_name) {
      return(RegulatoryDocument$new(drug_name = drug_name, document_type = "Biologics Regulatory Document"))
    }
  )
)

process_document <- function(factory, drug_name) {
  document <- factory$create_document(drug_name)
  document$show_document()
}

vaccine_factory <- VaccineDocumentFactory$new()
process_document(vaccine_factory, "Test Vaccine 01")

small_molecule_factory <- SmallMoleculeDocumentFactory$new()
process_document(small_molecule_factory, "Test Small Molecule 01")

biologics_factory <- BiologicsDocumentFactory$new()
process_document(biologics_factory, "Test Monoclonal Antibody 01")

File: Pharmaceutical_Company_Example/Creational/Java/Factory_Method.java
// Product
class RegulatoryDocument {
    private String drugName;
    private String documentType;

    public RegulatoryDocument(String drugName, String documentType) {
        this.drugName = drugName;
        this.documentType = documentType;
    }

    public void showDocument() {
        System.out.println("Regulatory Document for: " + drugName);
        System.out.println("Type of Document: " + documentType);
        System.out.println();
    }
}

// Abstract Creator (Factory)
abstract class RegulatoryDocumentFactory {
    public abstract RegulatoryDocument createDocument(String drugName);
}

// Concrete Creator for Vaccine Documents
class VaccineDocumentFactory extends RegulatoryDocumentFactory {
    public RegulatoryDocument createDocument(String drugName) {
        return new RegulatoryDocument(drugName, "Vaccine Regulatory Document");
    }
}

// Concrete Creator for Small-Molecule Documents
class SmallMoleculeDocumentFactory extends RegulatoryDocumentFactory {
    public RegulatoryDocument createDocument(String drugName) {
        return new RegulatoryDocument(drugName, "Small-Molecule Regulatory Document");
    }
}

// Concrete Creator for Biologics Documents
class BiologicsDocumentFactory extends RegulatoryDocumentFactory {
    public RegulatoryDocument createDocument(String drugName) {
        return new RegulatoryDocument(drugName, "Biologics Regulatory Document");
    }
}

// Client Code
public class Main {
    public static void processDocument(RegulatoryDocumentFactory factory, String drugName) {
        RegulatoryDocument document = factory.createDocument(drugName);
        document.showDocument();
    }

    public static void main(String[] args) {
        // Vaccine Regulatory Document
        RegulatoryDocumentFactory vaccineFactory = new VaccineDocumentFactory();
        processDocument(vaccineFactory, "Test Vaccine 01");

        // Small-Molecule Regulatory Document
        RegulatoryDocumentFactory smallMoleculeFactory = new SmallMoleculeDocumentFactory();
        processDocument(smallMoleculeFactory, "Test Small Molecule 01");

        // Biologics Regulatory Document
        RegulatoryDocumentFactory biologicsFactory = new BiologicsDocumentFactory();
        processDocument(biologicsFactory, "Test Monoclonal Antibody 01");
    }
}
